tes-adapter/src/main/java/com/epam/pipeline/tesadapter/service/TaskMapper.java

```diff
@@ -65,8 +80,54 @@ public PipelineStart mapToPipelineStart(TesTask tesTask) {
     }
 
     private TesExecutor getExecutorFromTesExecutorsList(List<TesExecutor> tesExecutors) {
-        Assert.notEmpty(tesExecutors, messageHelper.getMessage(MessageConstants.ERROR_PARAMETER_NULL_OR_EMPTY,
-                "executors"));
+        Assert.isTrue(tesExecutors.size() == ONLY_ONE, messageHelper.getMessage(
+                MessageConstants.ERROR_PARAMETER_INCOMPATIBLE_CONTENT, EXECUTORS, tesExecutors));
         return tesExecutors.get(FIRST);
     }
+
+    private String getProperInstanceType(TesTask tesTask, Tool pipelineTool) {
+        Double ramGb = tesTask.getResources().getRamGb();
+        Long cpuCores = tesTask.getResources().getCpuCores();
+        Long toolId = pipelineTool.getId();
+        Long regionId = getProperRegionIdInCloudRegionsByTesZone(tesTask.getResources().getZones());
+        Boolean spot = tesTask.getResources().getPreemptible();
+
+        AllowedInstanceAndPriceTypes allowedInstanceAndPriceTypes = cloudPipelineAPIClient
+                .loadAllowedInstanceAndPriceTypes(toolId, regionId, spot);
+        Assert.isTrue(allowedInstanceAndPriceTypes.getAllowedInstanceTypes() != null, messageHelper.getMessage(
+                MessageConstants.ERROR_PARAMETER_NULL_OR_EMPTY, allowedInstanceAndPriceTypes));
+        return evaluateMostProperInstanceType(allowedInstanceAndPriceTypes, ramGb, cpuCores);
+    }
+
+    private Tool loadToolByTesImage(String image) {
+        return cloudPipelineAPIClient.loadTool(image);
+    }
+
+    private Long getProperRegionIdInCloudRegionsByTesZone(List<String> zones) {
+        Assert.isTrue(zones.size() == ONLY_ONE, messageHelper.getMessage(
+                MessageConstants.ERROR_PARAMETER_INCOMPATIBLE_CONTENT, ZONES, zones));
+        return cloudPipelineAPIClient.loadAllRegions().stream().filter(
+                region -> region.getName().equalsIgnoreCase(zones.get(FIRST)))
+                .collect(Collectors.toList()).get(FIRST).getId();
+    }
+
+    private String evaluateMostProperInstanceType(AllowedInstanceAndPriceTypes allowedInstanceAndPriceTypes,
+                                                  Double ramGb, Long cpuCores) {
+        return allowedInstanceAndPriceTypes.getAllowedInstanceTypes().stream()
+                .collect(Collectors.toMap(InstanceType::getName, instancaType ->
+                        calculateInstanceCoef(instancaType, ramGb, cpuCores))).entrySet().stream()
+                .min(Comparator.comparing(Map.Entry::getValue)).orElseThrow(NullPointerException::new).getKey();
+    }
+
+    private Double calculateInstanceCoef(InstanceType instanceType, Double ramGb, Long cpuCores) {
```

Comment:
Also it's always a good practice to provide some comments with explanation of what this coefficient for and what a value shows

Reply:
fixed, added java-docs